Require explicit pool size in `pool_memory_resource` and move some things out of detail namespace

harrism opened the pull request:
## Description
Fixes #1416. 

 - ~Deprecates existing ctors of `pool_memory_resource` that provide optional parameter for the initial pool size.~
 - Adds new ctors that require an explicit initial pool size.
 - We don't yet deprecate anything in this PR because that would break builds of some RAPIDS libraries. We will follow up with PRs to cuDF, cuGraph and anything else needed to remove deprecated usages after this PR is merged.
 - Adds a new utility `fraction_of_available_device_memory` that calculates the specified fraction of free memory on the current CUDA device. This is now used in tests to provide an explicit pool size and can be used to produce the previous behavior of `pool_memory_resource` for consumers of the library.
 - Moves `available_device_memory` from a detail header to `cuda_device.hpp` so it is now publicly usable, along with the above utility.
 - Temporarily adds `detail::available_device_memory` as an alias of the above in order to keep cudf and cugraph building until we can update them.
 - Duplicates commonly externally used alignment functions that are currently in `rmm::detail` to the public `rmm` namespace.  The detail versions will be removed after cuDF and cuGraph are updated to not use them.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

harrism commented:
Discovering the cudf and cugraph depend on `rmm::detail::available_device_memory` so should probably add an alias for that rather than just moving it to `rmm::available_device_memory`. Then once the dependent libs are updated we can remove the detail version.

harrism commented:
/ok to test

harrism commented:
/ok to test

miscco reviewed (approved):
The changes look good to me.

The one thing that I find interesting is that this removes usage of  `get_upstream()->get_mem_info(cuda_stream_legacy)`

I am wondering whether we should deprecate that facility too

wence- commented on benchmarks/device_uvector/device_uvector_bench.cu:
```diff
@@ -38,7 +39,8 @@
   rmm::mr::cuda_memory_resource cuda_mr{};
-  rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource> mr{&cuda_mr};
+  rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource> mr{
+    &cuda_mr, rmm::percent_of_free_device_memory(1. / 2)};
```
**question** This is asking for an initial pool size of `0%` since there is no `rmm::percent_of_free_device_memory(double fraction)` and the numeric promotion turns 0.5 into 0 when resolving to `rmm::percent_of_free_device_memory(int percent)`.

Is that what you want, or did you mean:

```suggestion
    &cuda_mr, rmm::percent_of_free_device_memory(50)};
```

harrism replied:
Yes, this is a vestige of a past version. Thanks for catching.

wence- commented on include/rmm/cuda_device.hpp:
```diff
@@ -102,6 +103,52 @@ inline int get_num_cuda_devices()
+ *
+ * @return The available and total device memory in bytes for the current device as a std::pair.
+ */
+inline std::pair<std::size_t, std::size_t> available_device_memory()
```
**suggestion/nit** At point of use of this function, the user must "just remember" that the first entry in the pair is the free and the second is the total device memory. How about instead:

```
struct mem_info {
    std::size_t free;
    std::size_t total;
}

inline mem_info available_device_memory()
{
   ...
}
```
Instead?

harrism replied:
Often one ignores one of the return values e.g. with 

```
auto [[maybe_unused]] [free, _] = available_device_memory();
```

Does the struct approach allow this?

Aside: I always hated that cudaMemGetInfo returns two things. There should be two separate functions.

wence- replied:
Looks like it does, yes https://gcc.godbolt.org/z/eqzaG8cY9

wence- replied:
I don't think this is a blocking suggestion, so I'll let you choose whether or not to make this change, approving as is.

harrism replied:
Thanks.

wence- commented on include/rmm/cuda_device.hpp:
```diff
@@ -102,6 +103,52 @@ inline int get_num_cuda_devices()
+ * @brief Returns the approximate specified percent of free device memory on the current CUDA
+ * device, aligned to the nearest CUDA allocation size.
+ *
+ * @param percent The percent of free memory to return. Defaults to 50%.
```
**issue** This docstring is at odds with the implementation, which returns the requested percentage of _total_ device memory (or the remaining free memory if that number is too large). I think we want the implementation as is, but then docstring wants changed.

Perhaps

> Return the approximate specified percent of total device memory on the current CUDA device or the total free device memory (whichever is smaller), ...

?

harrism replied:
Good catch, thanks.

wence- commented on include/rmm/cuda_device.hpp:
```diff
@@ -102,6 +103,52 @@ inline int get_num_cuda_devices()
+{
+  auto const [free, total] = rmm::available_device_memory();
+
+  double fraction = static_cast<double>(percent) / 100;
```
**nit**: If we're going to the effort of casting `percent`:

```suggestion
  double fraction = static_cast<double>(percent) / 100.;
```

harrism commented:
Yes we should. :) See #1388 and #1389.

wence- commented on include/rmm/mr/device/pool_memory_resource.hpp:
```diff
@@ -110,6 +109,29 @@ class pool_memory_resource final
+   */
+  [[deprecated("Must specify initial_pool_size")]]  //
+  explicit pool_memory_resource(Upstream* upstream_mr,
+                                thrust::optional<std::size_t> initial_pool_size = thrust::nullopt,
+                                thrust::optional<std::size_t> maximum_pool_size = thrust::nullopt)
```
**question**: Why not `std::optional`? (since C++17), I guess for minimum code churn relative to the other constructors?

harrism replied:
Yeah we should switch them all in a separate PR. These ctors will be removed anyway.

wence- commented on include/rmm/mr/device/pool_memory_resource.hpp:
```diff
@@ -110,6 +109,29 @@ class pool_memory_resource final
+  }
+
   /**
    * @brief Construct a `pool_memory_resource` and allocate the initial device memory pool using
```
```suggestion
   * @brief [DEPRECATED] Construct a `pool_memory_resource` and allocate the initial device memory pool using
```

Although, **question**, does doxygen not pick up and mark `[[deprecated]]` objects in docs automatically?

harrism replied:
I think it does. I will add @deprecated to the docstring.

wence- commented on include/rmm/mr/device/pool_memory_resource.hpp:
```diff
@@ -286,28 +330,17 @@ class pool_memory_resource final
    * @param maximum_size The optional maximum size for the pool
    */
   // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
```
```suggestion
```

Parameters no longer easily swappable?

harrism replied:
Thanks.

wence- commented on include/rmm/mr/device/pool_memory_resource.hpp:
```diff
@@ -286,28 +330,17 @@ class pool_memory_resource final
    * @param maximum_size The optional maximum size for the pool
```
**issue**: This docstring no longer matches the function. Since `initial_size` can now no longer be optional and it doesn't query `get_mem_info`. Perhaps:

```
@brief Allocate initial memory for the pool

@param initial_size The initial size for the pool
@param maximum_size The optional maximum size

@throws logic_error if maximum_size is provided and is smaller than initial_size
```
?

wence- reviewed (changes requested):
Looks good, and I like making things more explicit. A few places where the docstrings are now at odds with the implementation.

harrism commented:
/ok to test

wence- commented:
I pushed a fix for the docs build (we need to include the new `deprecated` doxygen page in the rst docs for the links to resolve).

wence- commented on include/rmm/mr/device/pool_memory_resource.hpp:
```diff
@@ -110,32 +109,82 @@ class pool_memory_resource final
+   * @brief Construct a `pool_memory_resource` and allocate the initial device memory
+   * pool using `upstream_mr`.
+   *
+   * @deprecated Use the constructor that takes an explicit initial pool size instead.
```
It seems impossible to link to overloaded templated constructors in doxygen, so I was not able to do that.

wence- commented on python/docs/librmm_docs/deprecated.rst:
```diff
@@ -0,0 +1,5 @@
+Deprecated functionality
+========================
+
+.. doxygenpage:: deprecated
+   :content-only:
```
The addition of `@deprecated` tags produces a new page of doxygen output that we must include somewhere in the rst toctree such that sphinx will find the link anchors.

vuule commented on tests/mr/host/pinned_pool_mr_tests.cpp:
```diff
@@ -33,7 +32,7 @@ using pool_mr = rmm::mr::pool_memory_resource<rmm::mr::pinned_memory_resource>;
 TEST(PinnedPoolTest, ThrowOnNullUpstream)
 {
-  auto construct_nullptr = []() { pool_mr mr{nullptr}; };
+  auto construct_nullptr = []() { pool_mr mr{nullptr, 0}; };
```
is zero a valid pool size? If not, we should use a valid size to make sure the throw is because of the nullptr.

harrism replied:
Fixed anyway. Thanks!

wence- commented:
/ok to test

harrism commented:
Thanks for the doc fixes @wence-!  We shouldn't merge until I have PRs ready for RAPIDS repos that use the deprecated functions.

harrism commented:
/ok to test

harrism commented:
I have added a bunch to this PR. 
 - Removed deprecations. We should add replacement functionality first, then update dependent libraries, *then* deprecate.
 - Moved some detail functions out of detail that external libraries are currently using. We don't want external libraries using detail functions, but if they are externally useful, they shouldn't be in detail.

wence- commented on tests/mr/device/mr_ref_test.hpp:
```diff
@@ -17,10 +17,11 @@
 #pragma once
 
 #include "../../byte_literals.hpp"
+#include "rmm/cuda_device.hpp"
 
+#include <rmm/aligned.hpp>
```
```suggestion

#include <rmm/aligned.hpp>
#include <rmm/cuda_device.hpp>
```

wence- commented on tests/mr/device/mr_test.hpp:
```diff
@@ -17,10 +17,11 @@
 #pragma once
 
 #include "../../byte_literals.hpp"
+#include "rmm/cuda_device.hpp"
 
+#include <rmm/aligned.hpp>
```
```suggestion

#include <rmm/aligned.hpp>
#include <rmm/cuda_device.hpp>
```

wence- commented:
/ok to test

harrism commented on benchmarks/multi_stream_allocations/multi_stream_allocations_bench.cu:
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#include "rmm/cuda_device.hpp"
```
Oops these were auto-added.

harrism commented on include/rmm/aligned.hpp:
```diff
@@ -0,0 +1,96 @@
+ * @brief Returns whether or not `n` is a power of 2.
+ *
+ */
+constexpr bool is_pow2(std::size_t value) { return (0 == (value & (value - 1))); }
```
Right. And align_up(x, 0) is not very useful. Note due to clang-tidy complaints this will be fixed with the less satisfying `return (value != 0U) && ((value & (value - 1)) == 0U);`

Note TIL in C++20 we can replace `is_pow2` with `std::has_single_bit`

harrism commented:
/ok to test

harrism commented:
/ok to test

wence- commented:
/ok to test

wence- commented:
Figured out the problem with the cross-linking to objects defined in the `utilities` group (cc @vyars to check the `conf.py` changes), so now the docs build with correct linking.

So from my point of view this is definitely good to go.

vyasr commented:
The doc fixes look correct to me, thanks @wence-!

harrism commented:
/merge
